File: CanvasPetUnity/Assets/PetController.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.Runtime.InteropServices.WindowsRuntime;
using Unity.PlasticSCM.Editor.WebApi;
using Unity.VisualScripting;
using UnityEngine;
using UnityEngine.Assertions;

public class PetController : MonoBehaviour
{
    private Pet pet;
    private Queue<Notification> notificationQueue;
    private DateTime ScheduledUpdate;

    public NotificationWall notifWall;

    public int HAPPINESS_THRESHOLD = 750;

    private int HEALTH_CHECK_INTERVAL = 30;




    // Start is called before the first frame update
    void Start()
    {
        
        notificationQueue = new Queue<Notification>();

        //pull existing pet if exists
        pet = new Pet();

        if (pet == null)
        {
           pet = InitialPetSetup();
        }

    }

    // Update is called once per frame
    void Update()
    {
        //TESTING
        Notification notif = new AutoFeedNotification(pet, 9, "testing");
        notificationQueue.Enqueue(notif);
        //if (pet != null) {}

    }
    private void FixedUpdate()
    {
        if (notificationQueue != null)
        {
            Notify();
        }

        if (DateTime.Now >  ScheduledUpdate)
        {
            ScheduledUpdate = HealthCheck();
        }

       // if(pet.IsHungry())

    }

    int AutoFeed(int incFood, string assignmentName)
    {       
        if (pet.GetCurrentFood() < pet.MAX_FOOD)
        {
            int diff = pet.MAX_FOOD - pet.GetCurrentFood();
            if (diff > incFood) 
            {
                pet.AddFood(incFood);

                notificationQueue.Enqueue(new AutoFeedNotification(pet, incFood, assignmentName));
                return incFood;
            }
            else
            {
                pet.AddFood(diff);
                pet.StoreFood(incFood-diff);
                notificationQueue.Enqueue(new AutoFeedNotification(pet, incFood, incFood - diff, assignmentName));
                return diff;      
            }
        }
        else
        {
            pet.StoreFood(incFood);
            notificationQueue.Enqueue(new StorageNotification(pet, incFood, assignmentName));
            return 0;
        }
    }


    public void ForceUpdate()
    {
        pet?.ForceUpdate();
    }

    public void Notify()
    {
        if (notificationQueue != null && notificationQueue.Count > 0)
        {
            Notification notif = notificationQueue.Peek();
            if (notif != null && !notif.announced)
            {
                notifWall.AddNotification(notif);
            }
            notificationQueue.Dequeue();
        }
    }

    public DateTime HealthCheck()
    {
        DateTime currentTime = DateTime.Now;
        notifWall.SetHungerFlag(pet.IsHungry());
        notifWall.SetSadFlag(pet.CalculateHappiness()>HAPPINESS_THRESHOLD);


        return currentTime.AddMinutes(HEALTH_CHECK_INTERVAL);
    }

    

    public Pet InitialPetSetup()
    {
        Pet petTemp = new Pet();
        petTemp.SetBirthday(DateTime.Now);
            //extra stuff

        
        
        
        
        return petTemp;
    }
}

File: CanvasPetUnity/Assets/PetController.ts
import { AutoFeedNotification, Notification, StorageNotification } from "./Notification";
import { Pet } from "./Pet";
import { Queue } from "~lib/Queue";
import { NotificationWall } from "./NotificationWall";

const HAPPINESS_THRESHOLD: number = 800;

class PetController {
    private pet: Pet;
    private notificationQueue: Queue<Notification>;
    private scheduledUpdate: Date;
    public notifWall: NotificationWall;

    constructor() {
        this.pet = new Pet();
        this.notificationQueue = new Queue<Notification>();
        this.scheduledUpdate = new Date();
    
    }
    // Start is called before the first frame update
    start(): void {
        this.pet = new Pet();
        this.notificationQueue = new Queue<Notification>();
        this.pet.setState("idle"); // Set pet state to idle for testing
    }

    // Placeholder update method for idle state
    isIdle(): boolean {
        return true; // Always returns true for testing purposes
    }

    // Additional methods kept for uniformity with other group members
    update(): void {
        const notif = new AutoFeedNotification(this.pet, 9, "testing");
        this.notificationQueue.enqueue(notif);
    }

    fixedUpdate(): void {
        if (this.notificationQueue) {
            this.notify();
        }

        if (new Date() > this.scheduledUpdate) {
            this.scheduledUpdate = this.healthCheck();
        }
    }

    autoFeed(incFood: number, assignmentName: string): number {
        if (this.pet.getCurrentFood() < this.pet.MAX_FOOD) {
            const diff = this.pet.MAX_FOOD - this.pet.getCurrentFood();
            if (diff > incFood) {
                this.pet.addFood(incFood);
                this.notificationQueue.enqueue(new AutoFeedNotification(this.pet, incFood, assignmentName));
                return incFood;
            } else {
                this.pet.addFood(diff);
                this.pet.storeFood(incFood - diff);
                this.notificationQueue.enqueue(new AutoFeedNotification(this.pet, incFood, incFood - diff, assignmentName));
                return diff;
            }
        } else {
            this.pet.storeFood(incFood);
            this.notificationQueue.enqueue(new StorageNotification(this.pet, incFood, assignmentName));
            return 0;
        }
    }

    forceUpdate(): void {
        this.pet?.forceUpdate();
    }

    notify(): void {
        if (this.notificationQueue && this.notificationQueue.size() > 0) {
            const notif = this.notificationQueue.peek();
            if (notif && !notif.announced) {
                // this.notifWall.addNotification(notif);
            }
            this.notificationQueue.dequeue();
        }
    }

    healthCheck(): Date {
        const currentTime = new Date();
        this.notifWall.SetHungerFlag(this.pet.isHungry());
        this.notifWall.SetSadFlag(this.pet.calculateHappiness()>HAPPINESS_THRESHOLD);
        return new Date(currentTime.getTime() + 30 * 60000); // Add 30 minutes
        
    }
}

export default PetController;
